File: shared-libs/go/utils/health/health.go
// Package health provides /healthz and /readyz HTTP handlers for Go services.
package health

import (
	"encoding/json"
	"net/http"
)

// ReadinessCheck is a function that returns true if the service is ready, false otherwise.
type ReadinessCheck func() (ready bool, reason string)

// HealthzHandler responds 200 for liveness checks.
func HealthzHandler(w http.ResponseWriter, _ *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusOK)
	_ = json.NewEncoder(w).Encode(map[string]string{"status": "ok"})
}

// ReadyzHandler runs registered readiness checks.
func ReadyzHandler(checks ...ReadinessCheck) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		for _, check := range checks {
			if ok, reason := check(); !ok {
				w.WriteHeader(http.StatusServiceUnavailable)
				_ = json.NewEncoder(w).Encode(map[string]string{"status": "not_ready", "reason": reason})
				return
			}
		}
		w.WriteHeader(http.StatusOK)
		_ = json.NewEncoder(w).Encode(map[string]string{"status": "ready"})
	}
}

// MountHealthChecks registers /healthz and /readyz on a ServeMux.
func MountHealthChecks(mux *http.ServeMux, checks ...ReadinessCheck) {
	mux.HandleFunc("/healthz", HealthzHandler)
	mux.HandleFunc("/readyz", ReadyzHandler(checks...))
}

File: shared-libs/node/src/health.js
'use strict';

/**
 * Express health check middleware factory.
 * Mounts /healthz (liveness) and /readyz (readiness) on an Express app.
 */

/**
 * Adds /healthz and /readyz routes to an Express app.
 * @param {object} app - Express app instance
 * @param {Function} [readinessCheck] - Optional async fn returning {ready, reason}
 */
function mountHealthChecks(app, readinessCheck) {
  app.get('/healthz', (_req, res) => res.json({ status: 'ok' }));

  app.get('/readyz', async (_req, res) => {
    if (!readinessCheck) return res.json({ status: 'ready' });
    try {
      const result = await readinessCheck();
      if (result && result.ready === false) {
        return res.status(503).json({ status: 'not_ready', reason: result.reason });
      }
      res.json({ status: 'ready' });
    } catch (err) {
      res.status(503).json({ status: 'not_ready', reason: err.message });
    }
  });
}

module.exports = { mountHealthChecks };
